glommio/src/executor/mod.rs

```diff
@@ -1490,7 +1511,28 @@ impl LocalExecutor {
                     }
                 }
             }
-        })
+        };
+
+        #[cfg(not(feature = "native-tls"))]
+        {
+            assert!(
+                !LOCAL_EX.is_set(),
+                "There is already an LocalExecutor running on this thread"
+            );
+            LOCAL_EX.set(self, || run(self))
+        }
+
+        #[cfg(feature = "native-tls")]
+        unsafe {
+            assert!(
+                LOCAL_EX.is_null(),
+                "There is already an LocalExecutor running on this thread"
+            );
+
+            defer!(LOCAL_EX = std::ptr::null());
```

Comment:
I am concerned with a destruction race here.
When we call drop on the executor, we will then call drop on its task queues, which then call drop on tasks, etc.

Some of that code in drop may want to reference the reactor to issue async cancellations and the like, and now that will be null.

So I think the safest way would be for us to destroy the executor manually, and only then make it null.

If you think that couldn't happen, we'd have to prove that this is the case and write a big comment. It should be relatively easy to test too

Reply:
This code is written to exactly match the behavior of the other implementation (set the TLS var here and unset as soon as we return). The defer ensures we unset the tls var if we panic.

This was done this way because the std-tls are scoped and can't be freely accessed. This is not a limitation that I have with the native TLS, but I mirrored the behavior to avoid creating subtle differences in behavior between the two implementations.

So long story short this is not a problem and I couldn't implement your solution using the current TLS support on stable.